Upgrade to support `eslint@v9`

phun-ky opened the pull request:
* chore!: Upgrade to , and update eslint setup

What's cooking:

* Use typescript
* Update lint config
* use node native test runner
* use rollup
* upgraded to eslint v9
* got linting to run

Missing:

* Get tests to run/update tests
* Add generated documentation with typedoc 


✅ Closes: #947 #1015 #1083

Mettbrot commented:
@phun-ky can you take this up again? What is missing to get eslint upgraded?

phun-ky commented:
@Mettbrot I can take a look at it tomorrow and see if there's anything I can do more. I have some spare time moving forwards apaprently

Mettbrot commented:
Wow, cool Thanks! We’d need to find a maintainer of this, it seems pretty silent around the PRs  https://github.com/prettier/prettier-eslint/pull/1107

phun-ky commented:
@JounQin first blocker: `prettier-eslint-cli` uses `"eslint": "^8.51.0",` as a dep, and also `"@prettier/eslint": "npm:prettier-eslint@^16.1.0"` as a dep, and that dep uses `"eslint": "^8.7.0",` as a dep. I have updated to latest, but none of them can be used until they are fixed.

Mettbrot commented:
They live in this very repository too, right? Can we update them too and release them in one go?

JounQin commented:
`overrides` for `npm`/`pnpm` and `resolutions` for `yarn` could be used?

phun-ky commented:
hm, I think I've looked at this wrongly. The error I get is from v9: `Error: This method cannot be used with flat config. Add your entries directly into the config array.` Because the class `Linter` is used, and the input it expects is the old config. Which I find very odd tbh. I'm trying to find decent enough migration documentation with usage of the `Linter` in v9, with no luck atm

phun-ky commented:
I've posted this in the eslint discord:

> I have an issue with migrating prettier-eslint to eslint@v9. prettier-eslint uses the Linter , and for some obscure reasoning, I am not allowed to pass in a flat config (which is the new config in v9) into it, since assertEslintrcConfig is run , and it throws an error telling me that flat configs are not allowed.. Isn't every usage of eslint v9 requiring a flat config?

phun-ky commented:
I'm getting a bit closer. Latest update: converted to ts, added rollup. The things that remains: decide if we need to find a `getRules` equivalent: https://eslint.org/docs/latest/use/migrate-to-9.0.0#-linter-now-expects-flat-config-format, or to only use the rules passed in the options, and to get the linting and tests to run. Please feel free to work on this PR @JounQin

phun-ky commented:
Updated description to this PR. Getting closer. I am commenting out the rules in our own `eslint.config.mjs` for `import/no-unused-modules` until https://github.com/import-js/eslint-plugin-import/issues/3079 is fixed.

JounQin commented:
> Is that how it is supposed to be?

Yes, test files are also linted otherwise it should be listed in eslintignore.

phun-ky commented:
@JounQin nono, I meant, that the config passed to ESLint in the tests are merged with the ESLint config in the project root. Is that how it should work?

JounQin commented:
Yes, that's how eslint works, test files linting merges and overrides the root, so that we only enable/disable specific rules for test files.

phun-ky commented:
So to be clear, when testing, the passed config from the tests should override root, and if no passed config in the test exists, the root take presedence?

phun-ky commented:
hm, I see that the current tests has a flaw, or the program it self. with `filePath` in the test: https://github.com/prettier/prettier-eslint/blob/master/src/__tests__/index.js#L48, shipped to `analyze`,  we run this from ESLint: `calculateConfigForFile`. This function gets the closes relevant ESLint config file from the DIRECTORY of `filePath`, not using the rules/config defined in that file. Is this an oversight?

JounQin commented:
What is `eslint.default.config.mjs`? That would not be automatically loaded AFAIK.

phun-ky commented:
the equivalent of this: https://github.com/prettier/prettier-eslint/blob/master/src/__tests__/index.js#L57

phun-ky commented:
Let me see if I get this correctly:

If you use `format` with:

**1. no `filePath` and no config passed**

1. The nearest `eslint.config.mjs` of `cwd` (or any other supported config file) is found and used.

**2. no `filePath` and config passed**

1. The nearest `eslint.config.mjs` of `cwd` (or any other supported config file) is found and used.
2. The config passed is merged into the nearest `eslint.config.mjs` found.

**3. `filePath` given and no config passed**

1. The nearest `eslint.config.mjs` to `filePath` (or any other supported config file) is found and used.

**4. `filePath` given and config passed**

1. The nearest `eslint.config.mjs` to `filePath` (or any other supported config file) is found and used.
2. The config passed is merged into the nearest `eslint.config.mjs` found.

Does this seem correct? Or should, when config is given/passed (programtically and not from `eslint.config.mjs`, replace any found config?

phun-ky commented:
@JounQin do you have a chance to look at the PR and see if it's viable? I still need someone to take a look at the tests

phun-ky commented:
@JounQin still having issues with the config. please check out the message history here: https://discord.com/channels/688543509199716507/1336023233454936226/1343542887463260212

dvakatsiienko commented:
@phun-ky hi! do you need any help in this PR? what is left to do to take sure it works and merge it?

phun-ky commented:
yeah, need to get the tests to work tbh..  @dvakatsiienko

dvakatsiienko commented:
Okay. no guarantees, but I'll try to find some time and take a look during this or next weekend.
I need this ext to work :)

JounQin commented:
Sorry for the delay, I'd work on this in my next schedule.

JounQin commented:
This PR is too big, I'll do it myself, sorry.
